### ISSUE TITLE
Add AES-CTR to Parquet Encryption 

### ISSUE THREAD

**Issue by asfimport** [apache/parquet-java#1566]
AES-GCM-CTR perform GCM encryption on metadata and CTR encryption on data.

AES-CTR would perform CTR encryption on both. 

During Perf testing, we found AES-CTR can improve read/write performance by ~10% comparing with AES-GCM-CTR.

 

I checked with Gidon and the initial assumption was that AES-GCM-CTR would have similar performance as AES-CTR. But with recent performance benchmarking, we found it is worthy to introduce AES-CTR. Since many companies strive for parquet performance improvement. 

 

**Reporter**: [Xinli Shang](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=shangx@uber.com) / @shangxinli
#### PRs and other links:
- [GitHub Pull Request #153](https://github.com/apache/parquet-format/pull/153)

<sub>**Note**: *This issue was originally created as [PARQUET-1659](https://issues.apache.org/jira/browse/PARQUET-1659). Please see the [migration documentation](https://issues.apache.org/jira/browse/PARQUET-2502) for further details.*</sub>

**Comment by asfimport** [comment]
[Gidon Gershinsky](https://issues.apache.org/jira/browse/PARQUET-1659?#comment-16934945) / @ggershinsky:
A number of comments
1. The idea behind AES_GCM_CTR is that in Parquet files the metada size is negligible compared to the data size. E.g., the default page size is around a megabyte, and the page header is say a hundred bytes - meaning the metadata is around 0.01% of data. When running with old Java (without AES-NI acceleration), GCM is indeed slower than CTR, but applying it on 0.01% of the file shouldn't be noticeable.
1. Basic math shows that getting 10% speed up means that in your files the metadata part is around a few percents, instead of 0.01% - hundreds times larger. Would be good to analyze the reasons (small pages? or small files where the footer size becomes a considerable part of  the file size? anything else?) - this is easy to measure.
1. Java 9 and later run AES in hardware, so its possible to get full data protection (both encryption and integrity guarantees) via GCM without noticeable change in throughput.
1. Introduction of a third algorithm means additions in the spec and in the thrift, which could require another round of PMC vote. We might want to proceed with the parquet-2.7.0 release as is, and in parallel investigate the reasons of the 10% speed up in your files (item 2). If we decide to add a pure CTR algo, it can be a part of say parquet-2.7.1.

**Comment by asfimport** [comment]
[Xinli Shang](https://issues.apache.org/jira/browse/PARQUET-1659?#comment-16935553) / @shangxinli:
Yes, I understood the philosophy of designing GCM_CTR. The page size is default 1M. The footer size shouldn't matter because I commented out the code to generate the signature for the footer.  

What I suspect is that there could be some **base** overhead for invoking GCM even you have very few data(metadata in this case) to generate the signature. 

Many companies are still using Java 8 I think. 

Introducing **CTR only** doesn't mean rule out **GCM/GCM_CTR**. It just adds one more option for people to choose as a trade-off between performance and security. Unless **CTR  only** itself has a security hole, which I don't think so, it would be nice to let people choose. 

Yes, that would require to change the spec. But this is a very lightweight change and it doesn't change any direction of the original encryption design. I am not sure it still require PMC votes(any spec change need PMC vote?), but I think once we have agreement on the technical part, we can bring to PMCs to ask.

**Comment by asfimport** [comment]
[Gidon Gershinsky](https://issues.apache.org/jira/browse/PARQUET-1659?#comment-16935610) / @ggershinsky:
Sounds good, we are basically in sync on this. Lets understand the technical details behind the observed difference. For each encrypted module type (pages, page headers, columnmetadata, etc) could you find the number of modules and the total size of each type in your files (by simply adding a debug code in the right places).

**Comment by asfimport** [comment]
[Gidon Gershinsky](https://issues.apache.org/jira/browse/PARQUET-1659?#comment-16935626) / @ggershinsky:
lets discuss these numbers via mail, or google doc, etc, to spare the dev mailing list subscribers.

**Comment by asfimport** [comment]
[Antoine Pitrou](https://issues.apache.org/jira/browse/PARQUET-1659?#comment-17311611) / @pitrou:
Closing as this hasn't generated any further interest. Feel free to open a new issue with additional data if desired.
